src/endpoints/authentication/impl.cpp

```diff
@@ -20,120 +21,484 @@
-            res.set(field_type::server, irods::http::version::server_name);
-            res.set(field_type::content_type, "text/plain");
-            res.keep_alive(_req.keep_alive());
-            res.body() = std::move(bearer_token);
-            res.prepare_payload();
-
-            return _sess_ptr->send(std::move(res));
-        });
-    } // authentication
-} // namespace irods::http::endpoint
+	auto hit_token_endpoint(std::string encoded_body) -> nlohmann::json
+	{
+		const auto token_endpoint{
+			irods::http::globals::oidc_endpoint_configuration().at("token_endpoint").get<const std::string>()};
+
+		// Setup net
+		net::io_context io_ctx;
+		net::ip::tcp::resolver tcp_res{io_ctx};
+		beast::tcp_stream tcp_stream{io_ctx};
+
+		// Setup curl
+		CURLU* endpoint{curl_url()};
+
+		// Parse url
+		CURLUcode rc{curl_url_set(endpoint, CURLUPART_URL, token_endpoint.data(), 0)};
+		if (rc != 0) {
+			log::debug("Something happend....");
+		}
+
+		// Get host
+		char* host{};
+		rc = curl_url_get(endpoint, CURLUPART_HOST, &host, 0);
+		if (rc != 0) {
+			log::debug("Something happend....");
+		}
+
+		// Get service/port
+		//char *port{};
+		//rc = curl_url_get(endpoint, CURLUPART_PORT, &port, 0);
+		//if (rc != 0) {
+		//    log::debug("Something happend....");
+		//}
+		// KEYCLOAK does not return the port?
+		const auto port{irods::http::globals::oidc_configuration().at("port").get<const std::string>()};
+
+		// Get path
+		char* path{};
+		rc = curl_url_get(endpoint, CURLUPART_PATH, &path, 0);
+		if (rc != 0) {
+			log::debug("Something happend....");
+		}
+
+		// Addr
+		const auto resolve{tcp_res.resolve(host, port)};
+
+		// TCP thing
+		tcp_stream.connect(resolve);
+
+		// Build Request
+		constexpr auto version_number{11};
+		beast::http::request<beast::http::string_body> req{beast::http::verb::post, path, version_number};
+		req.set(beast::http::field::host, host);
+		req.set(beast::http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+		req.set(beast::http::field::content_type,
+		        "application/x-www-form-urlencoded"); // Possibly set a diff way?
+
+		// Send
+		req.body() = encoded_body;
+		req.prepare_payload();
+
+		// Send request
+		beast::http::write(tcp_stream, req);
+
+		// Read back req
+		beast::flat_buffer buffer;
+		beast::http::response<beast::http::string_body> res;
+		beast::http::read(tcp_stream, buffer, res);
+
+		log::debug("Got the following resp back: {}", res.body());
+
+		// Close socket
+		beast::error_code ec;
+		tcp_stream.socket().shutdown(net::ip::tcp::socket::shutdown_both, ec);
+
+		// Free up all items created, reverse n all
+		curl_free(path);
+		//curl_free(port);
+		curl_free(host);
+
+		// Done
+		curl_url_cleanup(endpoint);
+
+		// JSONize response
+		return nlohmann::json::parse(res.body());
+	}
+
+	auto encode_string(std::string_view to_encode) -> std::string
+	{
+		// Init CURL
+		CURL* curl{curl_easy_init()};
+
+		// Ensure that CURL was successfully enabled
+		if (curl != nullptr) {
```

Comment:
Turns out we don't need `curl_easy_init()` so we can remove it and the `nullptr` check.

Thinking of moving the `encode_string(...)` function to common, since it may be useful in other areas of the code base.

Reply:
Moved func to `encode(...)` in common. Forgot to prefix function arguments with `_` though...